wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParser.java

```diff
@@ -103,50 +112,80 @@
       }
       collectedFQNs.add(pFQN);
     }
     return collectedFQNs;
   }
 
   public ExtendedPluginFQN parsePluginFQN(String plugin) throws InfrastructureException {
-    String registry;
+    String url;
     String id;
     String publisher;
     String name;
     String version;
     URI registryURI = null;
     Matcher matcher = PLUGIN_PATTERN.matcher(plugin);
     if (matcher.matches()) {
-      registry = matcher.group("registry");
+      url = matcher.group("url");
       id = matcher.group("id");
       publisher = matcher.group("publisher");
       name = matcher.group("name");
       version = matcher.group("version");
     } else {
       throw new InfrastructureException(format(INCORRECT_PLUGIN_FORMAT_TEMPLATE, plugin));
     }
-    if (!isNullOrEmpty(registry)) {
-      try {
-        registryURI = new URI(registry);
-      } catch (URISyntaxException e) {
-        throw new InfrastructureException(
-            format(
-                "Plugin registry URL '%s' is invalid. Problematic plugin entry: '%s'",
-                registry, plugin));
+    if (!isNullOrEmpty(url)) {
+      if (isNullOrEmpty(id)) {
+        // reference only
+        return new ExtendedPluginFQN(url);
+      } else {
+        // registry + id
+        try {
+          registryURI = new URI(url);
+        } catch (URISyntaxException e) {
+          throw new InfrastructureException(
+              format(
+                  "Plugin registry URL '%s' is invalid. Problematic plugin entry: '%s'",
+                  url, plugin));
+        }
       }
     }
-
     return new ExtendedPluginFQN(registryURI, id, publisher, name, version);
   }
 
   private String[] splitAttribute(String attribute) {
     String[] plugins = attribute.split(",");
     return Arrays.stream(plugins).map(String::trim).toArray(String[]::new);
   }
+
+  /**
+   * Evaluates plugin FQN from provided reference by trying to fetch and parse its meta information.
+   *
+   * @param reference plugin reference to evaluate FQN from
+   * @param fileContentProvider content provider instance to perform plugin meta requests
+   * @return plugin FQN evaluated from given reference
+   * @throws InfrastructureException if plugin reference is invalid or inaccessible
+   */
+  public ExtendedPluginFQN evaluateFqn(String reference, FileContentProvider fileContentProvider)
+      throws InfrastructureException {
+    try {
+      String pluginMetaContent = fileContentProvider.fetchContent(reference);
+      JsonNode contentNode = yamlReader.readTree(pluginMetaContent);
+      String publisher = contentNode.get("publisher").textValue();
+      String name = contentNode.get("name").textValue();
+      String version = contentNode.get("version").textValue();
+      return new ExtendedPluginFQN(reference, publisher, name, version);
+    } catch (DevfileException | IOException e) {
+      throw new InfrastructureException(format("Plugin reference URL '%s' is invalid.", reference));
```

Comment:
Please propagate exception message here.
Will be an error message clear if JSON does not contain required nodes?

Reply:
Good point, made it more NPE-proof.